src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -216,15 +216,15 @@ protected override void Dispose(bool disposing)
             {
                 _disposed = true;
 
-                if (disposing)
+                // TODO (Issue 2508): Pinned buffers must be released in the callback, when it is guaranteed no further
+                // operations will be made to the send/receive buffers.
+                if (_cachedReceivePinnedBuffer.IsAllocated)
                 {
-                    // TODO (Issue 2508): Pinned buffers must be released in the callback, when it is guaranteed no further
-                    // operations will be made to the send/receive buffers.
-                    if (_cachedReceivePinnedBuffer.IsAllocated)
-                    {
-                        _cachedReceivePinnedBuffer.Free();
-                    }
+                    _cachedReceivePinnedBuffer.Free();
```

Comment:
just for good measure, consider setting _cachedReceivePinnedBuffer here to default(GCHandle).